src/std/misc/atom.ss

```diff
@@ -0,0 +1,88 @@
+;; Atoms in the style of clojure: shared, synchronous, independent state
+
+;; NB: Compared to the Clojure atom API, we:
+;; - Do not support the meta field and the functions alter-meta! reset-meta!
+;; - Replace swap-vals! and reset-vals! by swap-values! and reset-values!
+;;   returning values rather than vectors.
+;; - Rename add-watch and remove-watch to add-watch! and remove-watch!
+;; - Do not have generic functions that work on multiple kinds of "references"
+;;   but only one set of functions for atoms.
+;; - Also export atom-increment! and atomic-counter for the common case.
+
+(import :gerbil/gambit/threads
+        :std/sugar
+        :std/values)
+
+(export atom deref
+        swap! swap-values! reset! reset-values!
+        compare-and-set!
+        get-validator set-validator!
+        add-watch! remove-watch!
+        atom-increment! atomic-counter)
+
+(defstruct Atom
+  (mutex value validator watchers)
+  final: #t)
```

Comment:
could also add `unchecked: #t` to get unchecked accessors/mutators for internal use.